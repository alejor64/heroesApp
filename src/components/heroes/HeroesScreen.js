import React, { useMemo } from 'react'
import { Redirect, useParams } from 'react-router-dom'
import { getHeroesById } from '../../selector/getHeroesById'

const HeroesScreen = ({history}) => {

    const {heroeID} = useParams()
    const heroe = useMemo(() => getHeroesById(heroeID), [heroeID])

    if(!heroe){
        return <Redirect to="/" />
    }

    const {superhero, publisher, alter_ego, first_appearance, characters} = heroe

    const handleReturn = () => {
        if(history.length <= 2 && publisher==='DC Comics'){
            history.push('/dc')
        }else if (history.length <= 2 && publisher==='Marvel Comics') {
            history.push('/marvel')
        } else {
            history.goBack()
        }
    }

    return (
        <div className="row mt-5">
            <div className="col-4">
                <img
                    src={`../assets/heroes/${heroeID}.jpg`}
                    className="img-thumbnail animate__animated animate__fadeInLeft"
                    alt={superhero}
                />
            </div>
            <div className="col-8 animate__animated animate__fadeIn">
                <h3>{superhero}</h3>
                <ul className="list-group list-group-flush">
                    <li className="list-group-item">
                        <b>Alter ego: {alter_ego}</b>
                    </li>
                    <li className="list-group-item">
                        <b>Publisher: {publisher}</b>
                    </li>
                    <li className="list-group-item">
                        <b>First Apparance: {first_appearance}</b>
                    </li>
                </ul>

                <h5>Characters</h5>
                <p>{characters}</p>

                <button
                    className="btn btn-outline-info"
                    onClick={handleReturn}
                >
                    Return
                </button>
            </div>
        </div>
    )
}

export default HeroesScreen
