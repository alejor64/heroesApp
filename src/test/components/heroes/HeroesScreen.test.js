import React from 'react'
import '@testing-library/jest-dom'
import { mount } from 'enzyme'
import HeroesScreen from '../../../components/heroes/HeroesScreen'
import { MemoryRouter, Route } from 'react-router-dom'

describe('Pruebas en el componente <HeroesScreen />', () => {
    const historyMock = {
        length: 10,
        push: jest.fn(),
        goBack: jest.fn()
    }

    test('Debe de mostrarse el componente <Redirect /> si no hay argumentos en el URL', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/heroe']}>
                <HeroesScreen
                    history={historyMock}
                />
            </MemoryRouter>
        )

        expect(wrapper).toMatchSnapshot()
        expect(wrapper.find('Redirect').exists()).toBe(true)
    })
    
    test('Debe mostrar un heroe si el parÃ¡metro existe', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/heroe/marvel-hulk']}>
                <Route path="/heroe/:heroeID" component={HeroesScreen} />
            </MemoryRouter>
        )

        expect(wrapper.find('h5').text()).toBe('Characters')
        expect(wrapper.find('img').exists()).toBe(true)
    })

    test('Debe de regresar a la pantalla anterior con PUSH', () => {
        const historyMock = {
            length: 1,
            push: jest.fn(),
            goBack: jest.fn()
        }

        const wrapper = mount(
            <MemoryRouter initialEntries={['/heroe/marvel-hulk']}>
                <Route
                    path="/heroe/:heroeID"
                    component={() => <HeroesScreen history={historyMock} />}
                />
            </MemoryRouter>
        )

        wrapper.find('button').simulate('click')
        expect(historyMock.push).toHaveBeenCalledWith('/marvel')
        expect(historyMock.goBack).not.toHaveBeenCalled()
    })

    test('Debe de regresar a la pantalla anterior con GOBACK', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/heroe/marvel-hulk']}>
                <Route
                    path="/heroe/:heroeID"
                    component={() => <HeroesScreen history={historyMock} />}
                />
            </MemoryRouter>
        )

        wrapper.find('button').simulate('click')
        expect(historyMock.goBack).toHaveBeenCalled()
        expect(historyMock.push).not.toHaveBeenCalled()
    })
    
    test('Debe de llamar el <Redirect /> si el hero no existe', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/heroe/marvel-hulk123456']}>
                <Route
                    path="/heroe/:heroeID"
                    component={() => <HeroesScreen history={historyMock} />}
                />
            </MemoryRouter>
        )
        expect(wrapper.text()).toBe('')
    })
    
})
