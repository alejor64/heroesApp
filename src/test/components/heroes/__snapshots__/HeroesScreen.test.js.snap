// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <HeroesScreen /> Debe de mostrarse el componente <Redirect /> si no hay argumentos en el URL 1`] = `""`;
